### PR TITLE
pyqt@5: support python@3.10

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by danielnachun** [comment]
Similar to what we've done for other Python formulae with a lot of dependents, I'd like to see if we can refactor `pyqt@5` to support both `python@3.9` and `python@3.10`.  This will allow us to migrate the dependents of `pyqt@5` to `python@3.10` individually rather than in one large PR.  

The only snag here is that `pyqt@5` does provide 3 shell scripts in `bin` that are currently calling hardcoded to call `python3` from `python@3.10` (presumably because that is the last Python version used in the install).  I think we can just change this to be `python3`, under the assumption that dependents will make sure the PATH points to the right version.  I haven't made this change yet but will do so if there is agreement that this is the right solution.

File: Formula/pyqt@5.rb [hunk 1]
```diff
@@ -15,8 +16,9 @@ class PyqtAT5 < Formula
   end
 
   depends_on "pyqt-builder" => :build
+  depends_on "python@3.10"  => :build
+  depends_on "python@3.9"   => :build
```
Comment (carlocab) [inline Formula/pyqt@5.rb:20]: You need one of these to be available during the test, no? Preferably both of them so we can test both.

File: Formula/pyqt@5.rb [hunk 2]
```diff
@@ -61,27 +63,38 @@ class PyqtAT5 < Formula
...
-        system "sip-install", "--target-dir", site_packages
+      resource("PyQt5-sip").stage do
+        system python, *Language::Python.setup_install_args(prefix),
+                       "--install-lib=#{prefix/Language::Python.site_packages(python)}"
+      end
+
+      components.each do |p|
+        resource(p).stage do
+          inreplace "pyproject.toml", "[tool.sip.project]",
+            "[tool.sip.project]\nsip-include-dirs = [\"#{site_packages}/PyQt#{version.major}/bindings\"]\n"
```
Comment (carlocab) [inline Formula/pyqt@5.rb:95]:
```suggestion
          inreplace "pyproject.toml", "[tool.sip.project]", <<~EOS
            [tool.sip.project]
            sip-include-dirs = ["#{site_packages}/PyQt#{version.major}/bindings"]
          EOS
```
Something like this should work and is nicer to read.

File: Formula/pyqt@5.rb (hunk 1 above)
Comment (danielnachun) [reply Formula/pyqt@5.rb:20]: Whoops, they should definitely both be test dependencies.  The whole point of the test is try both!

**Comment by danielnachun** [comment]
I think everything is working now but because we have to compile twice, it will need `CI-long-timeout` to finish.

File: Formula/pyqt@5.rb [hunk 3]
```diff
@@ -61,37 +63,52 @@ class PyqtAT5 < Formula
...
+            [tool.sip.project]
+            sip-include-dirs = ["#{site_packages}/PyQt#{version.major}/bindings"]
+          EOS
+          system "sip-install", "--target-dir", site_packages
+        end
       end
     end
+
+    # Replace hardcoded reference to Python version used with sip/pyqt-builder with generic python3.
+    bin.children.each { |script| inreplace script, Formula["python@3.10"].opt_bin/"python3.10", "python3" }
```
Comment (carlocab) [inline Formula/pyqt@5.rb:104]: Are these shebangs? Or something else?

**Review by carlocab — APPROVED** [review]

File: Formula/pyqt@5.rb (hunk 3 above)
Comment (danielnachun) [reply Formula/pyqt@5.rb:104]: No, not shebangs, just a literal call to `python3`.  The scripts are just bash scripts.

**Comment by fxcoudert** [comment]
That `inreplace` does not actually work consistently, see https://github.com/Homebrew/homebrew-core/pull/115587

**Pull request closed without merging**